README.md
# Micro Monitor

[![Build Status](https://img.shields.io/travis/npm/micro-monitor/master.svg)](https://travis-ci.org/npm/micro-monitor)
[![Coverage Status](https://coveralls.io/repos/npm/micro-monitor/badge.svg?branch=master)](https://coveralls.io/r/npm/micro-monitor?branch=master)

Add a standardized monitoring endpoint to your application. Especially
useful for adding monitoring to services that do not expose an
HTTP interface.

## Basic Usage

Simply initialize the monitor when your application starts up:

```js
const Monitor = require('micro-monitor')

let monitor = Monitor(9999, () => {
// do something now that monitoring is running
})
```

* `http://0.0.0.1:9999/_monitor/status` is now available, and will
  respond with a `200` status and status object:

```json
{
  "pid": 42176,
  "uptime": 0.796,
  "rss": {
    "rss": 53907456,
    "heapTotal": 37728256,
    "heapUsed": 26032248
  },
  "cmdline": [
    "/Users/benjamincoe/.nvm/versions/node/v7.1.0/bin/node",
    "/Users/benjamincoe/npm-inc/micro-monitor/node_modules/mocha/bin/_mocha",
    "test.js"
  ],
  "git": "b0c57aa"
}
```

* `http://0.0.0.1:9999/_monitor/ping` is also available and will respond with
  status `200` and the text `pong`.

## Customizing Status Information

You can customize the status information returned using `contribute`:

* **monitor.contribute(contributor)**: contribute additional information
  to the status object returned on `/_monitor/status`.
  * `contributor`: a function returning the object to supplement the status
    object with.

## License

ISC


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="3.0.0"></a>
# [3.0.0](https://github.com/npm/micro-monitor/compare/v2.1.0...v3.0.0) (2017-03-27)


### Bug Fixes

* use raw HTTP, return full server object ([856b5a1](https://github.com/npm/micro-monitor/commit/856b5a1))


### BREAKING CHANGES

* no longer uses restify



<a name="2.1.0"></a>
# [2.1.0](https://github.com/npm/micro-monitor/compare/v2.0.1...v2.1.0) (2017-03-25)


### Features

* add promise support ([7b839b4](https://github.com/npm/micro-monitor/commit/7b839b4))



<a name="2.0.0"></a>
# 2.0.0 (2016-12-29)


### Bug Fixes

* whoops, had typoe in JSON ([869d461](https://github.com/npm/micro-monitor/commit/869d461))


### Features

* allow additional values to be "contributed" to the status object ([#2](https://github.com/npm/micro-monitor/issues/2)) ([4cd73dd](https://github.com/npm/micro-monitor/commit/4cd73dd))
* pull [@ceejbot](https://github.com/ceejbot)'s code into a module ([b0c57aa](https://github.com/npm/micro-monitor/commit/b0c57aa))


### BREAKING CHANGES

* switchd to @chrisdickinson's API suggestion.



<a name="1.1.0-0"></a>
# 1.1.0-0 (2016-12-29)


### Bug Fixes

* whoops, had typoe in JSON ([869d461](https://github.com/npm/micro-monitor/commit/869d461))


### Features

* allow the key on an arbitrary object to be monitored ([e5c8754](https://github.com/npm/micro-monitor/commit/e5c8754))
* pull [@ceejbot](https://github.com/ceejbot)'s code into a module ([b0c57aa](https://github.com/npm/micro-monitor/commit/b0c57aa))


test.js
/* global describe, it, before */

require('chai').should()

const Monitor = require('./')
const request = require('request')

describe('micro-monitor', () => {
  let monitor = null
  before((done) => { monitor = Monitor(9999, done) })

  it('responds with status information', (done) => {
    request.get({
      url: 'http://127.0.0.1:9999/_monitor/status',
      json: true
    }, (err, res, status) => {
      if (err) return done(err)
      res.statusCode.should.equal(200)
      // should return an object representing memory usage.
      ;(typeof status.rss).should.equal('object')
      return done()
    })
  })

  it('responds with ping', (done) => {
    request.get({
      url: 'http://127.0.0.1:9999/_monitor/ping',
      json: true
    }, (err, res, response) => {
      if (err) return done(err)
      res.statusCode.should.equal(200)
      response.should.equal('pong')
      return done()
    })
  })

  describe('contribute', () => {
    it('allows you to contribute new and exciting keys to the status object', (done) => {
      const obj = {batman: 'rich but grumpy'}
      monitor.contribute(() => {
        return {
          batman: obj.batman
        }
      })

      request.get({
        url: 'http://127.0.0.1:9999/_monitor/status',
        json: true
      }, (err, res, response) => {
        if (err) return done(err)
        monitor.contribute(function () {})
        res.statusCode.should.equal(200)
        response.batman.should.equal('rich but grumpy')
        return done()
      })
    })
  })

  describe('missing endpoint', () => {
    it('still 404s', (done) => {
      request.get({
        url: 'http://127.0.0.1:9999/_monitor/status-but-no-i-kid',
        json: true
      }, (err, res, response) => {
        if (err) return done(err)
        res.statusCode.should.equal(404)
        return done()
      })
    })
  })
})


index.js
'use strict'

const http = require('http')
const assert = require('assert')
const Promise = require('bluebird')
const exec = require('child_process').exec
const logger = require('bole')('monitor')

let contributor = null

module.exports = function startMonitor (port, callback) {
  const monitor = http.createServer()
  let commitHash = ''

  monitor.on('request', (req, res) => {
    if (req.url !== '/_monitor/ping') {
      return
    }
    req.handled = true
    res.end('pong')
  })

  monitor.on('request', (req, res) => {
    if (req.url !== '/_monitor/status') {
      return
    }
    req.handled = true
    const getExtra = (
      contributor
      ? Promise.try(() => contributor())
      : Promise.resolve({})
    )

    const getStatus = getExtra.then(obj => {
      return Object.assign({
        pid: process.pid,
        uptime: process.uptime(),
        rss: process.memoryUsage(),
        cmdline: process.argv,
        git: commitHash
      }, obj)
    })

    return getStatus.then(status => {
      res.writeHead(200, {'content-type': 'application/json'})
      res.end(JSON.stringify(status))
    })
  })

  monitor.on('request', (req, res) => {
    logger.info(req)
    if (req.handled) {
      return
    }
    res.writeHead(404)
    res.end('')
  })

  exec('git rev-parse --short HEAD', function (err, stdout, stderr) {
    if (!err && stdout) commitHash = stdout.trim()
    monitor.listen(port, callback)
  })

  return Object.assign(monitor, {
    contribute: (_contributor) => {
      assert(typeof _contributor === 'function', 'contributor must be a function')
      contributor = _contributor
    }
  })
}
